### PR TITLE
CFY-4936 improved maintenance mode

### PR DESCRIPTION
**Pull request by nirbir**

### CODE REVIEW

File: rest-service/manager_rest/resources_v2_1.py [hunk 1]
```diff
@@ -35,55 +35,73 @@ class MaintenanceMode(SecuredResource):
...
-                write_maintenance_state(MAINTENANCE_MODE_ACTIVE)
-                return {'status': MAINTENANCE_MODE_ACTIVE}
+            state = utils.read_json_file(maintenance_file_path)
+
+            if state['status'] == MAINTENANCE_MODE_ACTIVE:
+                return state
+            if state['status'] == ACTIVATING_MAINTENANCE_MODE:
+                running_executions = utils.get_running_executions()
+                if running_executions:
+                    state['remaining_executions'] = running_executions
```
Comment (nirbir) [inline rest-service/manager_rest/resources_v2_1.py]: change to remaining_running_executions

File: rest-service/manager_rest/resources_v2_1.py [hunk 2]
```diff
@@ -35,55 +35,73 @@ class MaintenanceMode(SecuredResource):
...
+                    return state
+
+                # No 'else' clause since this will never happen. If there are
+                # no running executions, maintenance mode would have
+                # been activated at the maintenance handler hook (server.py)
+
         else:
-            return {'status': NOT_IN_MAINTENANCE_MODE}
+            return _prepare_maintenance_dict(NOT_IN_MAINTENANCE_MODE)
```
Comment (nirbir) [inline rest-service/manager_rest/resources_v2_1.py]: change this and other constants to maintenance_mode_deactivated...

File: rest-service/manager_rest/resources_v2_1.py [hunk 3]
```diff
@@ -35,55 +35,73 @@ class MaintenanceMode(SecuredResource):
...
-
-                write_maintenance_state(MAINTENANCE_MODE_ACTIVE)
-                return {'status': MAINTENANCE_MODE_ACTIVE}
+            state = utils.read_json_file(maintenance_file_path)
+
+            if state['status'] == MAINTENANCE_MODE_ACTIVE:
+                return state
+            if state['status'] == ACTIVATING_MAINTENANCE_MODE:
+                running_executions = utils.get_running_executions()
+                if running_executions:
```
Comment (nirbir) [inline rest-service/manager_rest/resources_v2_1.py]: no need for if

File: rest-service/manager_rest/resources_v2_1.py [hunk 4]
```diff
@@ -35,55 +35,73 @@ class MaintenanceMode(SecuredResource):
...
 
         if maintenance_action == 'activate':
             if os.path.isfile(maintenance_file_path):
-                return {'status': MAINTENANCE_MODE_ACTIVE}, 304
-
+                state = utils.read_json_file(maintenance_file_path)
+                return state, 304
+
+            now = str(datetime.now())
+            user = ''
```
Comment (nirbir) [inline rest-service/manager_rest/resources_v2_1.py]: user = rest_security.get_username() or ''

File: rest-service/manager_rest/resources_v2_1.py [hunk 5]
```diff
@@ -35,55 +35,73 @@ class MaintenanceMode(SecuredResource):
...
+                    requested_by=user)
+            utils.write_dict_to_json_file(maintenance_file_path, new_state)
 
-            return {'status': ACTIVATING_MAINTENANCE_MODE}
+            return new_state
 
         if maintenance_action == 'deactivate':
             if not os.path.isfile(maintenance_file_path):
-                return {'status': NOT_IN_MAINTENANCE_MODE}, 304
+                return _prepare_maintenance_dict(NOT_IN_MAINTENANCE_MODE), 304
```
Comment (nirbir) [inline rest-service/manager_rest/resources_v2_1.py]: validate activate or deactivate
raise informative exception if not.

File: rest-service/manager_rest/server.py [hunk 6]
```diff
@@ -160,6 +161,11 @@ def log_response(response):
 
 def handle_maintenance_mode():
```
Comment (nirbir) [inline rest-service/manager_rest/server.py]: create module maintenance.py and move stuff there

File: rest-service/manager_rest/server.py [hunk 7]
```diff
@@ -160,6 +161,11 @@ def log_response(response):
...
 
+    # enabling internal requests
+    if utils.is_internal_request(request) \
```
Comment (nirbir) [inline rest-service/manager_rest/server.py]: refactor

File: rest-service/manager_rest/test/security/test_authorization.py [hunk 8]
```diff
@@ -503,6 +510,58 @@ def _test_list_node_instances(self):
         self._assert_unauthorized(self.simple_user_client.node_instances.list)
         return node_instances
 
+    ###########################
+    # maintenance mode methods
+    ###########################
+    # NOTE: update roles_config.yaml for every rest version supported.
+
+    def _test_get_status_maintenance_mode(self):
+        deactivating_status = 'deactivated'
```
Comment (nirbir) [inline rest-service/manager_rest/test/security/test_authorization.py]: deactivated

File: rest-service/manager_rest/test/security/test_authorization.py [hunk 9]
```diff
@@ -503,6 +510,58 @@ def _test_list_node_instances(self):
...
+
+        # ...but deployers, viewers and simple users should not
+        self._assert_unauthorized(
+                self.deployer_client.maintenance_mode.activate)
+        self._assert_unauthorized(self.viewer_client.maintenance_mode.activate)
+        self._assert_unauthorized(
+                self.simple_user_client.maintenance_mode.activate)
+
+    def _test_deactivate_maintenance_mode(self):
+        deactivating_status = 'deactivated'
```
Comment (nirbir) [inline rest-service/manager_rest/test/security/test_authorization.py]: deactivated

File: rest-service/manager_rest/utils.py [hunk 10]
```diff
@@ -142,3 +146,48 @@ def mkdirs(folder_path):
...
         else:
             raise
+
+
+def get_running_executions():
+    executions = get_blueprints_manager().executions_list(
+            is_include_system_workflows=True).items
+    running_executions = []
+    for execution in executions:
+        if execution.status not in models.Execution.END_STATES:
```
Comment (nirbir) [inline rest-service/manager_rest/utils.py]: add status to response

File: rest-service/manager_rest/utils.py [hunk 11]
```diff
@@ -142,3 +146,48 @@ def mkdirs(folder_path):
...
+                'deployment_id': execution.deployment_id,
+                'workflow_id': execution.workflow_id
+            })
+
+    return running_executions
+
+
+def read_json_file(file_path):
+    with open(file_path) as f:
+        json_data = json.load(f)
```
Comment (nirbir) [inline rest-service/manager_rest/utils.py]: just return

File: rest-service/manager_rest/test/test_maintenance_mode.py [hunk 12]
```diff
@@ -64,80 +79,84 @@ def test_request_approval_in_maintenance_mode(self):
...
     def test_multiple_maintenance_mode_activations(self):
         self._activate_maintenance_mode()
         try:
             self._activate_maintenance_mode()
             self.fail('Expected the second start request to fail '
                       'since maintenance mode is already started.')
-        except exceptions.CloudifyClientError as e:
+        except exceptions.NotModifiedError as e:
             self.assertEqual(304, e.status_code)
+        self.assertTrue('already activated' in e.message)
```
Comment (nirbir) [inline rest-service/manager_rest/test/test_maintenance_mode.py]: assertIn

File: rest-service/manager_rest/test/test_maintenance_mode.py [hunk 13]
```diff
@@ -64,80 +79,84 @@ def test_request_approval_in_maintenance_mode(self):
...
+
+    def test_transition_to_active(self):
+        execution = self._start_maintenance_transition_mode()
+        self._terminate_execution(execution.id)
```
Comment (nirbir) [inline rest-service/manager_rest/test/test_maintenance_mode.py]: assert activating

File: rest-service/manager_rest/test/test_maintenance_mode.py [hunk 14]
```diff
@@ -148,6 +167,55 @@ def test_request_approval_in_maintenance_transition_mode(self):
...
+        response = self.client.maintenance_mode.status()
+        self.assertIsNone(response.remaining_executions)
+
+    def test_execution_amount_maintenance_deactivated(self):
+        self._activate_and_deactivate_maintenance_mode()
+        response = self.client.maintenance_mode.status()
+        self.assertIsNone(response.remaining_executions)
+
+    def test_execution_amount_maintenance_activating(self):
+        execution_info = self._start_maintenance_transition_mode()
```
Comment (nirbir) [inline rest-service/manager_rest/test/test_maintenance_mode.py]: execution

File: rest-service/manager_rest/test/test_maintenance_mode.py [hunk 15]
```diff
@@ -148,6 +167,55 @@ def test_request_approval_in_maintenance_transition_mode(self):
...
+
+    def test_trigger_time_maintenance_activated(self):
+        self._activate_maintenance_mode()
+        response = self.client.maintenance_mode.status()
+        self.assertTrue(len(response.activated_at) > 0)
+
+    def test_trigger_time_maintenance_deactivated(self):
+        self._activate_and_deactivate_maintenance_mode()
+        response = self.client.maintenance_mode.status()
+        self.assertTrue(len(response.activated_at) == 0)
```
Comment (nirbir) [inline rest-service/manager_rest/test/test_maintenance_mode.py]: not consistent

File: rest-service/manager_rest/test/test_maintenance_mode.py [hunk 16]
```diff
@@ -148,6 +167,55 @@ def test_request_approval_in_maintenance_transition_mode(self):
...
+
+    def test_trigger_time_maintenance_activating(self):
+        self._start_maintenance_transition_mode()
+        response = self.client.maintenance_mode.status()
+        self.assertTrue(len(response.activation_requested_at) > 0)
+
+    def test_requested_by_secured(self):
```
Comment (nirbir) [inline rest-service/manager_rest/test/test_maintenance_mode.py]: there are secured unit tests

File: rest-service/manager_rest/test/test_maintenance_mode.py [hunk 17]
```diff
@@ -168,24 +236,31 @@ def test_multiple_maintenance_mode_deactivations(self):
             self.client.maintenance_mode.deactivate()
             self.fail('Expected the second stop request to fail '
                       'since maintenance mode is not active.')
-        except exceptions.CloudifyClientError as e:
+        except exceptions.NotModifiedError as e:
             self.assertEqual(304, e.status_code)
+        self.assertTrue('already deactivated' in e.message)
 
     def test_maintenance_file(self):
         maintenance_file = os.path.join(self.maintenance_mode_dir,
```
Comment (nirbir) [inline rest-service/manager_rest/test/test_maintenance_mode.py]: move to test_utils.py

File: rest-service/manager_rest/test/test_maintenance_mode.py [hunk 18]
```diff
@@ -208,7 +283,7 @@ def _activate_maintenance_mode(self):
 
     def _start_maintenance_transition_mode(self):
         (blueprint_id, deployment_id, blueprint_response,
-         deployment_response) = self.put_deployment('d1')
+         deployment_response) = self.put_deployment('transition')
```
Comment (nirbir) [inline rest-service/manager_rest/test/test_maintenance_mode.py]: test with all non ended statuses

File: rest-service/manager_rest/maintenance.py [hunk 19]
```diff
@@ -0,0 +1,190 @@
...
+                    return result
+
+        if check_allowed_endpoint(request_endpoint):
+            return
+
+        if state['status'] == MAINTENANCE_MODE_ACTIVATED:
+            return maintenance_mode_error()
+
+
+def handle_activating_mode(state, request, request_endpoint):
```
Comment (dankilman) [inline rest-service/manager_rest/maintenance.py]: _

File: rest-service/manager_rest/maintenance.py [hunk 20]
```diff
@@ -0,0 +1,190 @@
...
+                state = prepare_maintenance_dict(
+                        MAINTENANCE_MODE_ACTIVATED,
+                        activated_at=now,
+                        remaining_executions=None,
+                        requested_by=state['requested_by'],
+                        activation_requested_at=state[
+                            'activation_requested_at'])
+                utils.write_dict_to_json_file(maintenance_file, state)
+            else:
+                result = handle_activating_mode(
```
Comment (dankilman) [inline rest-service/manager_rest/maintenance.py]: return

File: rest-service/manager_rest/server.py [hunk 21]
```diff
@@ -144,6 +139,10 @@ def log_request():
             headers_pretty_print(request.headers)))
 
 
+def handle_maintenance_mode():
```
Comment (dankilman) [inline rest-service/manager_rest/server.py]: clean

File: rest-service/manager_rest/test/test_maintenance_mode.py [hunk 22]
```diff
@@ -79,26 +80,20 @@ def test_request_approval_in_maintenance_mode(self):
...
             self.assertRaises(exceptions.MaintenanceModeActiveError,
                               self.client.blueprints.list)
 
     def test_external_request_denial_in_maintenance_mode(self):
         self._activate_maintenance_mode()
 
-        with mock.patch('manager_rest.utils.is_bypass_maintenance_mode'):
-            self.assertRaises(exceptions.MaintenanceModeActiveError,
-                              self.client.blueprints.list)
+        bypass_maintenance_client = self.create_client(
```
Comment (dankilman) [inline rest-service/manager_rest/test/test_maintenance_mode.py]: function

File: rest-service/manager_rest/test/test_maintenance_mode.py [hunk 23]
```diff
@@ -178,50 +175,44 @@ def test_execution_amount_maintenance_deactivated(self):
...
+    def test_no_user_in_unsecured_maintenance_activation(self):
+        self._activate_maintenance_mode()
+        response = self.client.maintenance_mode.status()
+        self.assertFalse(response.requested_by)
 
     def test_trigger_time_maintenance_activated(self):
         self._activate_maintenance_mode()
         response = self.client.maintenance_mode.status()
-        self.assertTrue(len(response.activated_at) > 0)
+        self.assertIsNotNone(response.activated_at)
```
Comment (dankilman) [inline rest-service/manager_rest/test/test_maintenance_mode.py]: assertTrue

File: rest-service/manager_rest/test/test_maintenance_mode.py [hunk 24]
```diff
@@ -178,50 +175,44 @@ def test_execution_amount_maintenance_deactivated(self):
...
-                        new=self.mock_prepare_maintenance_dict):
-            self._activate_maintenance_mode()
-            response = self.client.maintenance_mode.status()
-            self.assertEqual(response.requested_by, 'mock user')
-
-    def test_requested_by_non_secured(self):
-        self._activate_maintenance_mode()
-        response = self.client.maintenance_mode.status()
-        self.assertTrue(len(response.requested_by) == 0)
+        self.assertIsNotNone(response.activation_requested_at)
```
Comment (dankilman) [inline rest-service/manager_rest/test/test_maintenance_mode.py]: also here

File: rest-service/manager_rest/test/test_maintenance_mode.py [hunk 25]
```diff
@@ -277,22 +246,58 @@ def test_maintenance_mode_activating_error_raised(self):
                           blueprint_id='b1',
                           deployment_id='d1')
 
+    def test_pending_execution_maintenance_activating_error_raised(self):
```
Comment (dankilman) [inline rest-service/manager_rest/test/test_maintenance_mode.py]: extract

File: rest-service/manager_rest/maintenance.py [hunk 26]
```diff
@@ -153,18 +148,30 @@ def get_running_executions():
     return running_executions
 
 
-def is_internal_request(request):
-    return 'REMOTE_ADDR' in request.headers.environ \
-           and 'HTTP_HOST' in request.headers.environ \
-           and request.headers.environ['REMOTE_ADDR'] == \
-               request.headers.environ['HTTP_HOST']
+def _is_internal_request():
+    remote_addr_header = _get_remote_addr_header()
```
Comment (dankilman) [inline rest-service/manager_rest/maintenance.py]: use headers

File: rest-service/manager_rest/maintenance.py [hunk 27]
```diff
@@ -153,18 +148,30 @@ def get_running_executions():
...
+        request.headers.environ.get(http_host_header)
+
+
+def _is_bypass_maintenance_mode():
+    bypass_maintenance_header = 'HTTP_X_BYPASS_MAINTENANCE'
+    return bypass_maintenance_header in request.headers.environ\
+        and request.headers.environ.get(bypass_maintenance_header)
+
+
+def _get_remote_addr_header():
```
Comment (dankilman) [inline rest-service/manager_rest/maintenance.py]: if possible change

File: rest-service/manager_rest/test/test_maintenance_mode.py [hunk 28]
```diff
@@ -83,17 +83,37 @@ def test_request_approval_in_maintenance_mode(self):
...
+
+    def test_bypass_maintenance_denial_in_maintenance_mode(self):
         self._activate_maintenance_mode()
 
-        bypass_maintenance_client = self.create_client(
-                headers={'X_BYPASS_MAINTENANCE': 'yes'})
-        with mock.patch('manager_rest.maintenance.is_internal_request'):
-            bypass_maintenance_client.blueprints.list()
+        internal_request_client = self.create_client(
+                headers={'X-bypass-maintenance': 'true'})
```
Comment (dankilman) [inline rest-service/manager_rest/test/test_maintenance_mode.py]: CAPS

File: rest-service/manager_rest/maintenance.py [hunk 29]
```diff
@@ -149,26 +149,24 @@ def get_running_executions():
 
 
 def _is_internal_request():
-    remote_addr_header = _get_remote_addr_header()
-    http_host_header = _get_http_host_header()
-    return remote_addr_header in request.headers.environ \
-        and http_host_header in request.headers.environ \
-        and request.headers.environ.get(remote_addr_header) == \
-        request.headers.environ.get(http_host_header)
+    remote_addr = _get_remote_addr()
```
Comment (dankilman) [inline rest-service/manager_rest/maintenance.py]:
``` python
remote_addr = _get_remote_addr()
http_host = _get_host()
return all([remote_addr, http_host, remote_add == http_host])
```

File: rest-service/manager_rest/test/test_maintenance_mode.py [hunk 30]
```diff
@@ -83,35 +83,29 @@ def test_request_approval_in_maintenance_mode(self):
...
+            with patch('manager_rest.maintenance._get_host',
+                       new=self._get_host):
+                internal_request_bypass_maintenance_client.blueprints.list()
 
     def test_bypass_maintenance_denial_in_maintenance_mode(self):
         self._activate_maintenance_mode()
 
         internal_request_client = self.create_client(
-                headers={'X-bypass-maintenance': 'true'})
+                headers={'X_BYPASS_MAINTENANCE': 'true'})
```
Comment (dankilman) [inline rest-service/manager_rest/test/test_maintenance_mode.py]: change

**Pull request merged**